### ISSUE TITLE
Please Improve Documentation of "ComplexMethod"

### ISSUE THREAD

**Issue by dominicoder** [arturbosch/detekt#1009]
## Expected Behavior
The "ComplexMethod" rule documentation should clearly explain what constitutes a "complex" method and what should be done to resolve the failure.

Something like:
> Complex methods use too many `if`, `apply`, `let`, `whatever` blocks. 

> threshold (default: 10)
> 
> Total number of occurrence of `if`, `apply`, `let`, `whatever` blocks allowed for a method

## Current Behavior
A detekt failure for this rule on a method that has zero `if` statements (what I assume would be "complexity" due to branching logic) outputs `ComplexMethod - 14/10`.

14/10? 14/10 ... what? This method has zero `if` statements - what are you counting?

The [documentation](https://arturbosch.github.io/detekt/complexity.html#complexmethod) for "ComplexMethod" reads:

> Complex methods are hard to understand and read.

_OK_.
> It might not be obvious what side-effect a complex method has.

_Sure_.
> Prefer splitting up complex methods into smaller methods that are in turn easier to understand.

_Uh huh_.
> Smaller methods can also be named much clearer which leads to improved readability of the code. 

_Right_.

Wait ... that's the end of the docs? No explanation of what exactly is meant by _complex_?

And then:
> Configuration options:
> threshold (default: 10)
> 
> MCC threshold for a method

Wait ... what is "MCC threshold for a method"?

I had to look at [the source code](https://github.com/arturbosch/detekt/blob/master/detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/ComplexMethod.kt) to see a reference to a "McCabe visitor" and then had to Google _that_ to find out it refers to [cyclomatic complexity](https://en.wikipedia.org/wiki/Cyclomatic_complexity) which refers to branching logic which I had zero of in my method, and then had to Google `McCabeVisitor` to find [_its_ source code](https://github.com/arturbosch/detekt/blob/master/detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/internal/McCabeVisitor.kt) which is not in this project to _finally_ figure out that this rule wants me to use fewer `let`s, `apply`s, etc.

I never would have expected that having many of these functions would be considered "complex".

## Context
The error output left me confused as to what was wrong with my method and what 14/10 meant.
The documentation for said rule is so vague and general as to be useless in explaining what it is checking for and what you can do to resolve the issue.

## Your Environment
* Version used: 1.0.0.RC7-3
* Operating System and version: Mac OS High Sierra
* Link to your project: Private, sorry, but shouldn't be necessary.

**Comment by ldramire** [comment]
I'd be happy to update the docs based on my findings from reading the source, but I lack context on to _why_ the Kotlin lib functions are factored into "complexity". 

I think most people using Kotlin would agree that:
```kotlin
val something = Something().apply {
  foo();
  bar();
  fuzz();
  buzz();
}
```
is _less_ complex (and idiomatic) than:
```kotlin
val something = Something()
something.foo();
something.bar();
something.fuzz();
something.buzz();
```

I'd argue that the current implementation is wrong or bugged. Maybe this is what you're getting at with adding a flag to explicitly check for them?

In any case, like I said, I can work on updating the docs so they better spell out _what_ is happening based on the current implementation. I'll see when I can carve out some time to do that.

**Comment by arturbosch** [comment]
Yes your totally right. When I wrote this rule (and it was one of the first five) I had in mind that when I started programming in Kotlin I used this `utility or helper functions` to nest control flow. 

Something like `?.let { doStuff(); if (it ....) {} } `. 
Here clearly `?.let` is used for nesting instead of a null check (`if (var != null) { if (car ...) }`).

That I want to say is: you can use `apply, let etc` for nesting or for concise code. Its up to the developer.
I agree that we should not report them on default, but maybe with a flag? @schalkms what do you think :) ?

**Comment by arturbosch** [comment]
I think the simplest heuristic which could work would be checking if the klambda body has one statement or a statement sequence?

**Comment by tosborne1215** [comment]
I stumbled on this issue for the same reason.
